File: load-tests/src/main/java/PostStore.java
import com.faunadb.client.FaunaClient;
import com.faunadb.client.types.Value;

import java.util.List;
import java.util.concurrent.ExecutionException;

import static com.faunadb.client.query.Language.Class;
import static com.faunadb.client.query.Language.*;

public class PostStore {
    private static final String CLASS_NAME = "posts";

    private FaunaClient client;

    public PostStore(FaunaClient client) {
        this.client = client;
    }


    public Post create(String title, List<String> tags) throws InterruptedException, ExecutionException {
        Value result =
            client.query(
                Select(
                    Value("data"),
                    Let("id", NewId()).in(
                        Create(
                            Ref(Class(CLASS_NAME), Var("id")),
                                Obj("data",
                                    Obj("id", Var("id"), "title", Value(title), "tags", Value(tags))
                                )
                        )
                    )
                )
            ).get();

        return toPost(result);
    }

    public Post read(String id) throws InterruptedException, ExecutionException {
        Value result =
            client.query(
                Select(
                    Value("data"),
                    Get(Ref(Class(CLASS_NAME), Value(id)))
                )
            ).get();

        return toPost(result);
    }

    public Post update(String id, String title, List<String> tags) throws InterruptedException, ExecutionException {
        Value result =
            client.query(
                Select(
                    Value("data"),
                    Update(
                        Ref(Class(CLASS_NAME), Value(id)),
                        Obj("data",
                            Obj("title", Value(title), "tags", Value(tags))
                        )
                    )
                )
            ).get();

        return toPost(result);
    }

    public Post delete(String id) throws InterruptedException, ExecutionException {
        Value result =
            client.query(
                Select(
                    Value("data"),
                    Delete(Ref(Class(CLASS_NAME), Value(id)))
                )
            ).get();

        return toPost(result);
    }

    private Post toPost(Value value) {
        return value.to(Post.class).get();
    }
}

File: load-tests/src/main/scala/PostStore.scala
import faunadb.FaunaClient
import faunadb.query._
import faunadb.values.Value

import scala.concurrent.ExecutionContext.Implicits._
import scala.concurrent.Future

object PostStore {
  val ClassName = "posts"
}

class PostStore(client: FaunaClient) {
  import PostStore._

  def create(title: String, tags: Seq[String]): Future[Post] = {
    val result: Future[Value] =
      client.query(
        Select(
          "data",
          Let(
            Seq("id" -> NewId()),
            Create(
              Ref(Class(ClassName), Var("id")),
              Obj("data" -> Obj("id" -> Var("id"), "title" -> title, "tags" -> tags))
            )
          )
        )
      )

    result.flatMap(toPost)

  }

  def read(id: String): Future[Post] = {
    val result: Future[Value] =
      client.query(
        Select(
          "data",
          Get(Ref(Class(ClassName), id))
        )
      )

    result.flatMap(toPost)
  }

  def update(id: String, title: String, tags: Seq[String]): Future[Post] = {
    val result: Future[Value] =
      client.query(
        Select(
          "data",
          Update(
            Ref(Class(ClassName), id),
            Obj("data" -> Obj("title" -> title, "tags" -> tags))
          )
        )
      )

    result.flatMap(toPost)
  }

  def delete(id: String): Future[Post] = {
    val result: Future[Value] =
      client.query(
        Select(
          "data",
          Delete(Ref(Class(ClassName), Value(id)))
        )
      )

    result.flatMap(toPost)
  }

  private def toPost(value: Value): Future[Post] = {
    Future(value.to[Post].get)
  }

}
